Remotion/Web/Development.Analyzers/RMWEB0001_WebStringAnalyzer.cs

```diff
@@ -0,0 +1,223 @@
+          symbols.StringJoinMethodSymbols,
+          s => SymbolEqualityComparer.Default.Equals (invocationOriginalDefinitionSymbol, s),
+          out var methodSymbol))
+      {
+        var argumentTypes = GetArgumentTypeSymbols (node, ctx.SemanticModel);
+        if (AnyWithResult (
+            argumentTypes,
+            a => SymbolEqualityComparer.Default.Equals (a, symbols.WebStringTypeSymbol)
+                 || SymbolEqualityComparer.Default.Equals (a, symbols.PlainTextStringTypeSymbol)
+                 || ctx.Compilation.ClassifyConversion (a, symbols.WebStringEnumerable).IsImplicit
+                 || ctx.Compilation.ClassifyConversion (a, symbols.PlainTextStringEnumerable).IsImplicit,
+            out var result))
+        {
+          var diagnostic = Diagnostic.Create (
+              DiagnosticDescriptor,
+              ctx.Node.GetLocation(),
+              methodSymbol.ToString(),
+              result.ToString(),
+              $"Encode the {result.ToString()} instances first.");
+          ctx.ReportDiagnostic (diagnostic);
+        }
+      }
+    }
+
+    private static bool AnyWithResult<TSymbol> (IEnumerable<TSymbol> methods, Func<TSymbol, bool> predicate, out TSymbol match)
+        where TSymbol : ISymbol
+    {
+      match = methods.FirstOrDefault (predicate)!;
+      return match != null;
+    }
+
+    private static ITypeSymbol? GetFirstArgumentTypeSymbol (InvocationExpressionSyntax node, SemanticModel model)
+    {
+      var firstArgumentExpression = node.ArgumentList.Arguments.First().Expression;
+      return model.GetTypeInfo (firstArgumentExpression).Type;
+    }
+
+    private static IReadOnlyCollection<ITypeSymbol> GetArgumentTypeSymbols (InvocationExpressionSyntax node, SemanticModel model)
+    {
+      return node.ArgumentList.Arguments.Select (s => model.GetTypeInfo (s.Expression).Type).OfType<ITypeSymbol>().ToList();
+    }
+
+    private static IReadOnlyCollection<IMethodSymbol> GetStringJoinMethodSymbols (INamedTypeSymbol stringSymbol)
+    {
+      return stringSymbol
+          .GetMembers ("Join")
+          .Cast<IMethodSymbol>()
+          .ToList();
+    }
+
+    private static IMethodSymbol GetHtmlTextWriterWriteObjectMethodSymbol (INamedTypeSymbol htmlTextWriterSymbol, INamedTypeSymbol? objectSymbol)
+    {
+      return htmlTextWriterSymbol
+          .GetMembers ("Write")
+          .Cast<IMethodSymbol>()
+          .Single (m => m.Parameters.Length == 1 && SymbolEqualityComparer.Default.Equals (m.Parameters.Single().Type, objectSymbol));
+    }
+
+    private static IMethodSymbol GetStringBuilderAppendObjectMethodSymbol (INamedTypeSymbol stringBuilderTypeSymbol, INamedTypeSymbol? objectSymbol)
+    {
+      return stringBuilderTypeSymbol
+          .GetMembers ("Append")
+          .Cast<IMethodSymbol>()
+          .Single (m => m.Parameters.Length == 1 && SymbolEqualityComparer.Default.Equals (m.Parameters.Single().Type, objectSymbol));
+    }
+
+    class SymbolContext
```

Comment:
Question: would this be a good usecase for record?

Reply:
it would, it would require the IsExternalInit class to be defined though. But I guess that would be alright as its just the following:
```csharp
namespace System.Runtime.CompilerServices
{
      internal static class IsExternalInit {}
}
```

Reply:
Is this because we reference netstandard2.0 isntead of net5.0?

Would switching to net5.0 be a good idea?

And I don't mind the empty internal declaration, especially if it makes the code better readable.

Reply:
I had some issues running the tests with net5.0 but I'll look into it.

Reply:
Thanks. And it's more of a "is this actually the correct way to do this" question.

The big problem with netstandard2.0 is that microsoft has started to deprecate the concent. modern c# feautres aren't officially supported. NRT isn't officially supported (or annoated), etc. This means I try to avoid netstandard2.0 when I can use net 5 or net6 instead, hoping to make things easier moving forward.

Reply:
I agree, but iirc, analyzers should target netstandard2.0 - might be deprecated now tho.

Reply:
What ever's the correct practice :)